docs/design-docs/08-package-variant.md

```diff
@@ -0,0 +1,270 @@
+the variant.
+
+### Package Context[^notimplemented]
+Every kpt package that is designated `--for-deployment` will contain a
+ConfigMap called `kptfile.kpt.dev`. Kpt (or Porch) will automatically add a
+key `name` to the ConfigMap data, with the value of the package name. This
+ConfigMap can then be used as input to functions in the Kpt function pipeline.
+
+This process holds true for package revisions created via the package variant
+controller as well. Additionally, the author of the `PackageVariant` resource
+can specify additional key-value pairs to insert into the package
+context.
+
+### KRM Function Calls[^notimplemented]
+TODO(johnbelamaric): describe adding a KRM function pipeline to the
+PackageVariant to allow arbitrary mutations
+- question: should these allow adding to the package pipeline a la `--save`? Is
+  that a *separate* pipeline?
+
+### Configuration Injection[^pdc]
+Adding values to the package context, or executing functions with their
+configuration listed in the `PackageVariant` works for values that are under
+control of the author of the `PackageVariant` resource. However, in more
+advanced use cases, we may need to specialize the package based upon other
+contextual information. This particularly comes into play when the user
+deploying the workload does not have direct control over the context in which
+it is being deployed. For example, one part of the organization may manage the
+infrastructure - that is, the cluster in which we are deploying the workload -
+and another part the actual workload. We would like to be able to pull in inputs
+specified by the infrastructure team automatically, based the cluster to which
+we are deploying the workload, or perhaps the region in which that cluster is
+deployed.
+
+To facilitate this, the package variant controller can "inject" configuration
+directly into the package. This means it will use information specific to this
+instance of the package to lookup a resource in the Porch cluster, and copy that
+information into the package. Of course, the package has to be ready to receive
+this information. So, there is a protocol for facilitating this dance:
+- Packages may contain resources annotated with `kpt.dev/config-injection`
+- Usually, these will also be `config.kubernetes.io/local-config` resources, as
+  they are likely just used by local functions as input. But this is not
+  mandatory.
+- The package variant controller will look for any resource in the Kubernetes
+  cluster matching the Group, Version, and Kind of the package resource, and
+  satisfying the *injection selector*.
+- The package variant controller will copy the `spec` field from the matching
+  in-cluster resource to the in-package resource.
+
+Note that because we are injecting data *from the Kubernetes cluster*, we can
+also monitor that data for changes. For each resource we inject, the package
+variant controller will establish a Kubernetes "watch" on the resource. A change
```

Comment:
Note that we probably want to initiate a watch on the GVK, not the specific issue, to reduce the number of watches. But it's up for debate (see https://github.com/nephio-project/nephio-controller-poc/issues/24 originally).